zerver/tests/test_messages.py

```diff
@@ -2047,3 +2049,89 @@ def test_delete_message_second_time(self):
+            email=self.example_email("hamlet"))
+        long_term_idle_user.last_active_message_id = UserMessage.objects.filter(
+            user_profile=long_term_idle_user).order_by(
+            '-message__pub_date')[0].message_id
+        long_term_idle_user.long_term_idle = True
+        long_term_idle_user.save()
+
+        def idle_users_ums():
+            # type: () -> List[UserMessage]
+            return list(UserMessage.objects.filter(
+                user_profile=long_term_idle_user).order_by('message__pub_date'))
+
+        # Test that add_missing_messages() in simplest case of adding a
+        # message for which UserMessage row doesn't exist for this user.
+        sent_message = send_fake_message('Test Message 1')
+        idle_user_msg_count = len(idle_users_ums())
+        self.assertNotEqual(idle_users_ums()[-1].message, sent_message)
+        add_missing_messages(long_term_idle_user)
+        self.assertEqual(len(idle_users_ums()), idle_user_msg_count + 1)
+        self.assertEqual(idle_users_ums()[-1].message, sent_message)
+
+        # Test that add_missing_messages() only adds messages that aren't
+        # already present in the UserMessage table. This test works on the
+        # fact that previous test just above this added a message but didn't
+        # updated the last_active_message_id field for the user.
+        sent_message = send_fake_message('Test Message 2')
+        idle_user_msg_count = len(idle_users_ums())
+        self.assertNotEqual(idle_users_ums()[-1].message, sent_message)
+        add_missing_messages(long_term_idle_user)
+        self.assertEqual(len(idle_users_ums()), idle_user_msg_count + 1)
+        self.assertEqual(idle_users_ums()[-1].message, sent_message)
+
+        # Test UserMessage rows are created correctly in case of stream
+        # Subscription was altered by admin while user was away.
+        sent_message_list = []
+        sent_message_list.append(send_fake_message('Test Message 3'))
+        # Alter subscription to stream by simply creating a log of the fact
+        # that the subscription was altered.
+        log_change_in_subscription(long_term_idle_user, 'subscription_deactivated')
+        send_fake_message('Test Message 4')
+        log_change_in_subscription(long_term_idle_user, 'subscription_activated')
+        sent_message_list.append(send_fake_message('Test Message 5'))
+        sent_message_list.reverse()
+        idle_user_msg_count = len(idle_users_ums())
+        idle_user_msg_list = idle_users_ums()
+        for sent_message in sent_message_list:
+            self.assertNotEqual(idle_user_msg_list.pop().message, sent_message)
+        add_missing_messages(long_term_idle_user)
+        self.assertEqual(len(idle_users_ums()), idle_user_msg_count + 2)
+        idle_user_msg_list = idle_users_ums()
+        for sent_message in sent_message_list:
+            self.assertEqual(idle_user_msg_list.pop().message, sent_message)
```

Comment:
I think these tests could benefit from asserting how many total database queries are done by the `add_missing_messages` function, since it seems like a place where we really want to avoid database queries in a loop.